### PR TITLE
[CRM-51] add new car logic and tests

### PR DESCRIPTION
**Pull request by magikabdul**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Commits are suqashed
- [x] PR and commit messages follow the style guidelines

### CODE REVIEW

File: backend/src/main/java/com/crm/service/impl/CarServiceImpl.java [hunk 1]
```diff
@@ -49,4 +54,32 @@ public CarResponse getCarByRegistrationNumber(final String registrationNumber) {
...
+                    .orElseGet(() -> CarTypeEntity.builder().name("Other").build());
+        } else {
+            carTypeEntity = carTypeRepository.findByNameIgnoreCase(nameOfCarTypeEntity)
+                    .orElseGet(() -> CarTypeEntity.builder().name(nameOfCarTypeEntity).build());
+        }
+
+        CarEntity newCar = carMapper.convertToEntity(carRequest);
+        newCar.setCarTypeEntity(carTypeEntity);
+
+        return carRepository.save(newCar);
```
Comment (aliwocha) [inline backend/src/main/java/com/crm/service/impl/CarServiceImpl.java:83]: powinniśmy zwrócić CarResponse, czyli jeszcze by wypadało zmapować ten obiekt na Dto (zobacz u mnie jak to zaimplementowałem)

File: backend/src/test-integration/java/com/crm/api/v1/CarControllerIT.java [hunk 2]
```diff
@@ -24,9 +38,31 @@
...
+
     @Autowired
     private MockMvc mvc;
 
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @BeforeEach
+    public void setUp() {
+        carRequest = new CarRequest();
```
Comment (aliwocha) [inline backend/src/test-integration/java/com/crm/api/v1/CarControllerIT.java:51]: to chyba można usunąć?

File: backend/src/test/java/com/crm/service/impl/CarServiceImplTest.java [hunk 3]
```diff
@@ -89,4 +101,89 @@ public void getCarByRegistrationNumberShouldThrowExceptionForInvalidRegistration
...
+        verify(carRepository, times(1)).findByRegistrationNumberIgnoreCase(existingRegistrationNumber);
+    }
+
+    @Test
+    public void shouldAddCarWithTypeOtherWhenCarTypeNotProvided() {
+        carRequest = CarRequest.builder()
+                .carTypeEntity(CarTypeEntity.builder().name("").build())
+                .build();
+
+        when(carTypeRepository.findByNameIgnoreCase("Other"))
```
Comment (aliwocha) [inline backend/src/test/java/com/crm/service/impl/CarServiceImplTest.java:150]: nie rozdzielaj whenów - lepiej linijka pod linijką, żeby było widać podział na given - when -then

File: backend/src/test-integration/java/com/crm/api/v1/CarControllerIT.java [hunk 4]
```diff
@@ -134,4 +170,353 @@ void shouldReturnBadResponseStatusWhenCallingCarWithBlankRegistrationNumber() th
         mvc.perform(get("/cars/search?license-plate="))
                 .andExpect(status().is(NOT_FOUND.value()));
     }
-}
\ No newline at end of file
+
+    @Test
+    @SneakyThrows
+    void shouldResponseWith_BedRequest_WhenBodyIsMissing() {
```
Comment (aliwocha) [inline backend/src/test-integration/java/com/crm/api/v1/CarControllerIT.java:176]: literówka - Bad (we wszystkich testach)

File: backend/src/test-integration/java/com/crm/api/v1/CarControllerIT.java [hunk 5]
```diff
@@ -134,4 +170,353 @@ void shouldReturnBadResponseStatusWhenCallingCarWithBlankRegistrationNumber() th
...
+        assertTrue(message.contains(ErrorDict.CAR_PRODUCTION_YEAR_INVALID));
+        assertTrue(message.contains(ErrorDict.CAR_MILEAGE_INVALID));
+        assertTrue(message.contains(ErrorDict.CAR_TYPE_INVALID));
+    }
+
+    @Test
+    @SneakyThrows
+    void shouldResponseWith_BedRequest_WhenVinIsBlank() {
+        carRequest.setVin("");
+        String body = objectMapper.writeValueAsString(carRequest);
```
Comment (aliwocha) [inline backend/src/test-integration/java/com/crm/api/v1/CarControllerIT.java:209]: można by tu dać final i wszędzie w pozostałych testach

File: backend/src/main/java/com/crm/exception/GlobalExceptionHandler.java [hunk 6]
```diff
@@ -30,12 +30,18 @@ public ErrorResponse handleCustomerNotFoundException(final CustomerNotFoundExcep
...
-    @ResponseStatus(HttpStatus.NOT_FOUND)  
+    @ResponseStatus(HttpStatus.NOT_FOUND)
     @ExceptionHandler(CarNotFoundException.class)
     public ErrorResponse handleCarNotFoundException(final CarNotFoundException e, final HttpServletRequest request) {
         return new ErrorResponse(getSource(request), List.of(e.getMessage()));
     }
 
+    @ResponseStatus(HttpStatus.CONFLICT)
+    @ExceptionHandler(CarException.class)
+    public ErrorResponse handleCarHandlingException(final CarException e, final HttpServletRequest request) {
```
Comment (SteelWro) [inline backend/src/main/java/com/crm/exception/GlobalExceptionHandler.java:41]: i think handleCarException method name is enough

**Review by SteelWro — APPROVED** [review]

File: backend/src/main/java/com/crm/api/v1/CarController.java [hunk 7]
```diff
@@ -49,12 +49,12 @@
 
     @PostMapping
     @ApiOperation(value = "Add new car", notes = "Add body \"CarRequest\" to add new car to database")
-    public ResponseEntity<?> addCar(@Valid @RequestBody final CarRequest carRequest) {
-        CarEntity car = carService.addCar(carRequest);
+    public ResponseEntity<Void> addCar(@Valid @RequestBody final CarRequest carRequest) {
```
[the remark this answers is not in the document]
Comment (magikabdul) [reply backend/src/main/java/com/crm/api/v1/CarController.java:52]: Zrobiłem jak swego czasu ustaliliśmy

File: backend/src/main/java/com/crm/exception/GlobalExceptionHandler.java (hunk 6 above)
Comment (magikabdul) [reply backend/src/main/java/com/crm/exception/GlobalExceptionHandler.java:41]: Fixed
U're right. :) You can say that again. :)
kur.... więcej czasu straciłem na znalezienie odpowiedniego idiomu. :) .... a i tak nie jest to ten co chciałem pierwotnie. 🥇

**Review by aliwocha — APPROVED** [review]

**Pull request merged**